server/src/helpers/transaction.ts

```diff
@@ -58,93 +60,173 @@ export async function getOperations(
+          continue;
+        }
+      }
+
+      operations.push(withdrawOperation);
+      
       continue;
     }
 
     if (isDepositEvent(event)) {
-      operations.push(
-        Operation.constructFromObject({
-          operation_identifier: new OperationIdentifier(operations.length),
-          type: OpType.DEPOSIT,
-          status: OperationStatus.SUCCESS,
-          account: new AccountIdentifier(data[0].toString()),
-          amount: new Amount((data[1] as u128).toBn().toString(), currency),
-        }),
-      );
-      continue;
+      const account = data[0].toString()
+      const amount = data[1] as u128
+      
+      if(!transactionFeeDepositMatched) {
+        const amountsMatch = amount.eq(transactionFeeAmount)
+        
+        if (amountsMatch) {
+          transactionFeeDepositMatched = true
+
+          const transactionFeeCreditOperation = Operation.constructFromObject({
+            operation_identifier: new OperationIdentifier(operations.length),
+            type: OpType.TRANSACTION_FEE_PAID,
+            status: opStatus,
+            account: new AccountIdentifier(account),
+            amount: new Amount(amount.toBn().toString(), currency),
+            related_operations: [new OperationIdentifier(transactionFeeDebitOperationIndex)],
+          })
+
+          operations.push(transactionFeeCreditOperation);
+          continue;
+        }
+      }
+
+      const depositOperation = Operation.constructFromObject({
+        operation_identifier: new OperationIdentifier(operations.length),
+        type: OpType.DEPOSIT,
+        status: opStatus,
+        account: new AccountIdentifier(account),
+        amount: new Amount(amount.toString(), currency),
+      })
+
+      operations.push(depositOperation);
     }
 
     if (isReservedEvent(event)) {
+      const account = data[0].toString()
+      const amount = (data[1] as u128).toBn()
+
       operations.push(
         Operation.constructFromObject({
           operation_identifier: new OperationIdentifier(operations.length),
           type: OpType.RESERVED,
-          status: OperationStatus.SUCCESS,
-          account: new AccountIdentifier(data[0].toString()),
-          amount: new Amount((data[1] as u128).toBn().neg().toString(), currency),
+          status: opStatus,
+          account: new AccountIdentifier(account),
+          amount: new Amount(amount.clone().neg().toString(), currency),
         }),
       );
       continue;
     }
 
     if (isUnreservedEvent(event)) {
+      const account = data[0].toString()
+      const amount = (data[1] as u128).toBn()
+
       operations.push(
         Operation.constructFromObject({
           operation_identifier: new OperationIdentifier(operations.length),
           type: OpType.UNRESERVED,
-          status: OperationStatus.SUCCESS,
-          account: new AccountIdentifier(data[0].toString()),
-          amount: new Amount((data[1] as u128).toBn().toString(), currency),
+          status: opStatus,
```

Comment:
I am confused why all of these were defaulting to `OperationStatus.SUCCESS`?

Reply:
Because if these events occur, it means that these operations were successful even if the transaction failed.

Reply:
Look at [this block](https://idea.gear-tech.io/explorer/1232806?node=wss%3A%2F%2Fdev-archive-staking.vara-network.io) as an example. Transfer transaction failed, but Deposit, Withdraw and TransactionFeePaid operations were successful. Because transaction fee was paid despite the fact that transfer failed

Reply:
@osipov-mit is there an easy way to distinguish for things like deposit whether it should be marked as successful since one of them is related to the transaction fee?

Reply:
It always should be marked as succesful becuase if it's not then there will be no such events in the block.

Reply:
Okay, makes sense!

Reply:
@osipov-mit updated here!

https://github.com/gear-foundation/integrations-rosetta-api/pull/50/commits/c709670d89cc5163d9b860f82e4ac60a18af1cdf